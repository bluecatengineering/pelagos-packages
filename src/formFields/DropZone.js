import {useCallback} from 'react';
import PropTypes from 'prop-types';
import {t} from '@bluecat/l10n.macro';
import {faFileAlt} from '@fortawesome/free-regular-svg-icons';
import {faCloudUploadAlt} from '@fortawesome/free-solid-svg-icons';

import LabelLine from '../components/LabelLine';
import SvgIcon from '../components/SvgIcon';

import FieldError from './FieldError';
import './DropZone.less';

const handleClick = (event) => {
	event.currentTarget.firstChild.click();
};

const handleDragOver = (event) => {
	event.preventDefault();
	const dt = event.dataTransfer;
	dt.effectAllowed = 'copy';
	dt.dropEffect = 'copy';
};

const handleDragEnter = (event) => {
	const relatedTarget = event.relatedTarget;
	const currentTarget = event.currentTarget;
	if (!relatedTarget || !currentTarget.contains(relatedTarget)) {
		currentTarget.classList.add('DropZone__content--active');
	}
};

const handleDragLeave = (event) => {
	const relatedTarget = event.relatedTarget;
	const currentTarget = event.currentTarget;
	if (!relatedTarget || !currentTarget.contains(relatedTarget)) {
		currentTarget.classList.remove('DropZone__content--active');
	}
};

/** Input component which accepts a dropped file. */
const DropZone = ({
	id,
	className,
	label,
	optional,
	emptyHint,
	editingHint,
	uploadedHint,
	fileName,
	error,
	editing,
	onDrop,
}) => {
	const readFile = useCallback(
		(file) => {
			if (file) {
				const reader = new FileReader();
				reader.onload = () => onDrop(file.name, reader.result);
				reader.readAsText(file);
			}
		},
		[onDrop]
	);
	const handleDrop = useCallback(
		(event) => {
			event.preventDefault();
			event.currentTarget.classList.remove('DropZone__content--active');
			const {items, files} = event.dataTransfer;
			readFile(items ? items[0].getAsFile() : files[0]);
		},
		[readFile]
	);
	const handleChange = useCallback(
		(event) => {
			readFile(event.target.files[0]);
		},
		[readFile]
	);
	return (
		<div className={`DropZone${className ? ` ${className}` : ''}`}>
			<LabelLine htmlFor={id} text={label} optional={optional && !fileName && !editing} />
			<div
				id={id}
				className={'DropZone__content' + (error ? ' DropZone__content--error' : '')}
				tabIndex={0}
				role="button"
				onClick={handleClick}
				onDragOver={handleDragOver}
				onDragEnter={handleDragEnter}
				onDragLeave={handleDragLeave}
				onDrop={handleDrop}
			>
				<input style={{display: 'none'}} type="file" onChange={handleChange} />
				{fileName ? (
					<div className="DropZone__uploaded">
						<div>{t`"${fileName}" attached.`}</div>
						<SvgIcon icon={faFileAlt} className="DropZone__icon DropZone__icon--uploaded" />
						<div className="DropZone__hint">{uploadedHint}</div>
					</div>
				) : editing ? (
					<div className="DropZone__uploaded">
						<div>{editingHint}</div>
						<SvgIcon icon={faFileAlt} className="DropZone__icon DropZone__icon--uploaded" />
						<div className="DropZone__hint">{uploadedHint}</div>
					</div>
				) : (
					<div className="DropZone__empty">
						<div className="DropZone__hint">{emptyHint}</div>
						<SvgIcon icon={faCloudUploadAlt} className="DropZone__icon DropZone__icon--empty" />
					</div>
				)}
			</div>
			<FieldError text={error} />
		</div>
	);
};

DropZone.propTypes = {
	/** The component id. */
	id: PropTypes.string.isRequired,
	/** The component class name(s). */
	className: PropTypes.string,
	/** The label text. */
	label: PropTypes.string,
	/** Whether the field is optional. */
	optional: PropTypes.bool,
	/** The text to display when no file has been selected. */
	emptyHint: PropTypes.string,
	/** The text to display when no file has been selected in edit mode. */
	editingHint: PropTypes.string,
	/** The text to display when a file has been selected. */
	uploadedHint: PropTypes.string,
	/** The name of the selected file. */
	fileName: PropTypes.string,
	/** The error text. */
	error: PropTypes.string,
	/** Whether the field is in edit mode (changes optional and empty messages). */
	editing: PropTypes.bool,
	/** Function invoked when a file is dropped. */
	onDrop: PropTypes.func,
};

export default DropZone;
