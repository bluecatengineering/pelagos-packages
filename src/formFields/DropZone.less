@import '../../less/focus';
@import '../../less/fonts';
@import '../../less/spacing';

.DropZone {
	margin-top: @sp-12;
}

.DropZone__content {
	@focus-visible();
	display: flex;
	flex-direction: column;
	align-items: stretch;
	height: 155px;
	padding: @sp-08 @sp-12;
	background-color: var(--field);
	border: none;
	border-bottom: 1px solid var(--border-strong);
	border-top-left-radius: 3px;
	border-top-right-radius: 3px;
	text-align: left;
	cursor: pointer;
	transition: background-color 0.15s ease-out, border-color 0.15s ease-out;

	&:hover {
		background-color: var(--field-hover);
	}
}

.DropZone__content--error {
	@focus-visible-error();
	outline: 2px solid var(--support-error);
	outline-offset: -2px;
}
.DropZone__content--active {
	background-color: var(--field-hover);
}

.DropZone__uploaded {
	align-items: center;
}

.DropZone__icon {
	color: var(--layer-accent);
	margin: @sp-12 0;
	transition: color 0.15s ease-out;

	.DropZone__content--active & {
		color: var(--interactive);
	}

	.DropZone__content:hover & {
		color: var(--layer-accent-hover);
	}
}
.DropZone__icon--uploaded {
	font-size: 50px;
}
.DropZone__icon--empty {
	font-size: 90px;
	align-self: center;
}

.DropZone__hint {
	@font-12-400();
	color: var(--text-placeholder);
}

@media print {
	.DropZone__content {
		background-color: transparent;
		border-color: transparent;
	}
}
