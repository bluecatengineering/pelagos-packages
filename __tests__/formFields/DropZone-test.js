import {shallow} from 'enzyme';

import DropZone from '../../src/formFields/DropZone';

jest.unmock('../../src/formFields/DropZone');

describe('DropZone', () => {
	describe('rendering', () => {
		it('renders expected elements when fileName is not set', () => {
			const wrapper = shallow(
				<DropZone
					id="test"
					className="TestClass"
					label="Test"
					optional
					emptyHint="Empty hint"
					helperText="Helper text"
				/>
			);
			expect(wrapper.getElement()).toMatchSnapshot();
		});

		it('renders expected elements when fileName is set', () => {
			const wrapper = shallow(<DropZone id="test" uploadedHint="Uploaded hint" fileName="test-file.txt" />);
			expect(wrapper.getElement()).toMatchSnapshot();
		});

		it('renders expected elements when editing is true', () => {
			const wrapper = shallow(
				<DropZone id="test" uploadedHint="Uploaded hint" editingHint="Editing hint" editing={true} />
			);
			expect(wrapper.getElement()).toMatchSnapshot();
		});

		it('renders expected elements when error is set', () => {
			const wrapper = shallow(
				<DropZone id="test" uploadedHint="Uploaded hint" fileName="test-file.txt" error="Drop error" />
			);
			expect(wrapper.getElement()).toMatchSnapshot();
		});
	});

	describe('behaviour', () => {
		it('calls onDrop when a file is dropped', () => {
			const onDrop = jest.fn();
			const preventDefault = jest.fn();
			const remove = jest.fn();
			const name = 'file.txt';
			const result = 'text';
			const file = {name};
			const readAsText = jest.fn();
			const reader = {result, readAsText};
			global.FileReader = jest.fn(() => reader);
			const wrapper = shallow(<DropZone id="test" onDrop={onDrop} />);

			wrapper.find('#test').prop('onDrop')({
				preventDefault,
				currentTarget: {classList: {remove}},
				dataTransfer: {files: [file]},
			});
			expect(preventDefault.mock.calls).toEqual([[]]);
			expect(remove.mock.calls).toEqual([['DropZone__content--active']]);
			expect(readAsText.mock.calls).toEqual([[file]]);
			expect(reader.onload).toEqual(expect.any(Function));
			reader.onload();
			expect(onDrop.mock.calls).toEqual([[name, result]]);
		});

		it('calls onDrop when a file is dropped and items is available', () => {
			const onDrop = jest.fn();
			const preventDefault = jest.fn();
			const remove = jest.fn();
			const getAsFile = jest.fn();
			const name = 'file.txt';
			const file = {name};
			const readAsText = jest.fn();
			const reader = {readAsText};
			global.FileReader = jest.fn(() => reader);
			getAsFile.mockReturnValue(file);
			const wrapper = shallow(<DropZone id="test" onDrop={onDrop} />);

			wrapper.find('#test').prop('onDrop')({
				preventDefault,
				currentTarget: {classList: {remove}},
				dataTransfer: {items: [{getAsFile}]},
			});
			expect(preventDefault.mock.calls).toEqual([[]]);
			expect(remove.mock.calls).toEqual([['DropZone__content--active']]);
			expect(readAsText.mock.calls).toEqual([[file]]);
		});

		it('calls onDrop when a file is selected', () => {
			const onDrop = jest.fn();
			const name = 'file.txt';
			const file = {name};
			const readAsText = jest.fn();
			const reader = {readAsText};
			global.FileReader = jest.fn(() => reader);
			const wrapper = shallow(<DropZone id="test" onDrop={onDrop} />);

			wrapper.find('input').prop('onChange')({
				target: {files: [file]},
			});
			expect(readAsText.mock.calls).toEqual([[file]]);
		});

		it('does not call onDrop when no file is selected', () => {
			const onDrop = jest.fn();
			const readAsText = jest.fn();
			const reader = {readAsText};
			global.FileReader = jest.fn(() => reader);
			const wrapper = shallow(<DropZone id="test" onDrop={onDrop} />);

			wrapper.find('input').prop('onChange')({
				target: {files: []},
			});
			expect(readAsText.mock.calls).toEqual([]);
		});

		it('clicks the input element when the container is clicked', () => {
			const click = jest.fn();
			const wrapper = shallow(<DropZone id="test" />);
			wrapper.find('#test').simulate('click', {currentTarget: {firstChild: {click}}});
			expect(click.mock.calls).toEqual([[]]);
		});

		it('calls preventDefault on dragover', () => {
			const preventDefault = jest.fn();
			const dataTransfer = {};
			const wrapper = shallow(<DropZone id="test" />);
			wrapper.find('#test').prop('onDragOver')({preventDefault, dataTransfer});
			expect(preventDefault.mock.calls).toEqual([[]]);
			expect(dataTransfer).toEqual({effectAllowed: 'copy', dropEffect: 'copy'});
		});

		it('adds a class on dragenter', () => {
			const relatedTarget = {};
			const contains = jest.fn();
			const add = jest.fn();
			const wrapper = shallow(<DropZone id="test" />);
			wrapper.find('#test').prop('onDragEnter')({relatedTarget, currentTarget: {contains, classList: {add}}});
			expect(contains.mock.calls).toEqual([[relatedTarget]]);
			expect(add.mock.calls).toEqual([['DropZone__content--active']]);
		});

		it('does not add a class on dragenter when contains returns true', () => {
			const relatedTarget = {};
			const contains = jest.fn().mockReturnValue(true);
			const add = jest.fn();
			const wrapper = shallow(<DropZone id="test" />);
			wrapper.find('#test').prop('onDragEnter')({relatedTarget, currentTarget: {contains, classList: {add}}});
			expect(contains.mock.calls).toEqual([[relatedTarget]]);
			expect(add.mock.calls).toEqual([]);
		});

		it('removes a class on dragleave', () => {
			const relatedTarget = {};
			const contains = jest.fn();
			const remove = jest.fn();
			const wrapper = shallow(<DropZone id="test" />);
			wrapper.find('#test').prop('onDragLeave')({relatedTarget, currentTarget: {contains, classList: {remove}}});
			expect(contains.mock.calls).toEqual([[relatedTarget]]);
			expect(remove.mock.calls).toEqual([['DropZone__content--active']]);
		});

		it('does not remove a class on dragleave when contains returns true', () => {
			const relatedTarget = {};
			const contains = jest.fn().mockReturnValue(true);
			const remove = jest.fn();
			const wrapper = shallow(<DropZone id="test" />);
			wrapper.find('#test').prop('onDragLeave')({relatedTarget, currentTarget: {contains, classList: {remove}}});
			expect(contains.mock.calls).toEqual([[relatedTarget]]);
			expect(remove.mock.calls).toEqual([]);
		});
	});
});
