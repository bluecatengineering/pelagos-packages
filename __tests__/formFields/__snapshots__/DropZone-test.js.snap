// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DropZone rendering renders expected elements when editing is true 1`] = `
<div
  className="DropZone"
>
  <LabelLine
    htmlFor="test"
  />
  <div
    className="DropZone__content"
    id="test"
    onClick={[Function]}
    onDragEnter={[Function]}
    onDragLeave={[Function]}
    onDragOver={[Function]}
    onDrop={[Function]}
    role="button"
    tabIndex={0}
  >
    <input
      onChange={[Function]}
      style={
        Object {
          "display": "none",
        }
      }
      type="file"
    />
    <div
      className="DropZone__uploaded"
    >
      <div>
        Editing hint
      </div>
      <SvgIcon
        className="DropZone__icon DropZone__icon--uploaded"
        icon={
          Object {
            "icon": Array [],
            "iconName": "file-alt",
            "prefix": "far",
          }
        }
      />
      <div
        className="DropZone__hint"
      >
        Uploaded hint
      </div>
    </div>
  </div>
  <FieldError />
</div>
`;

exports[`DropZone rendering renders expected elements when error is set 1`] = `
<div
  className="DropZone"
>
  <LabelLine
    htmlFor="test"
  />
  <div
    className="DropZone__content DropZone__content--error"
    id="test"
    onClick={[Function]}
    onDragEnter={[Function]}
    onDragLeave={[Function]}
    onDragOver={[Function]}
    onDrop={[Function]}
    role="button"
    tabIndex={0}
  >
    <input
      onChange={[Function]}
      style={
        Object {
          "display": "none",
        }
      }
      type="file"
    />
    <div
      className="DropZone__uploaded"
    >
      <div>
        "test-file.txt" attached.
      </div>
      <SvgIcon
        className="DropZone__icon DropZone__icon--uploaded"
        icon={
          Object {
            "icon": Array [],
            "iconName": "file-alt",
            "prefix": "far",
          }
        }
      />
      <div
        className="DropZone__hint"
      >
        Uploaded hint
      </div>
    </div>
  </div>
  <FieldError
    text="Drop error"
  />
</div>
`;

exports[`DropZone rendering renders expected elements when fileName is not set 1`] = `
<div
  className="DropZone TestClass"
>
  <LabelLine
    htmlFor="test"
    optional={true}
    text="Test"
  />
  <div
    className="DropZone__content"
    id="test"
    onClick={[Function]}
    onDragEnter={[Function]}
    onDragLeave={[Function]}
    onDragOver={[Function]}
    onDrop={[Function]}
    role="button"
    tabIndex={0}
  >
    <input
      onChange={[Function]}
      style={
        Object {
          "display": "none",
        }
      }
      type="file"
    />
    <div
      className="DropZone__empty"
    >
      <div
        className="DropZone__hint"
      >
        Empty hint
      </div>
      <SvgIcon
        className="DropZone__icon DropZone__icon--empty"
        icon={
          Object {
            "icon": Array [],
            "iconName": "cloud-upload-alt",
            "prefix": "fas",
          }
        }
      />
    </div>
  </div>
  <FieldError />
</div>
`;

exports[`DropZone rendering renders expected elements when fileName is set 1`] = `
<div
  className="DropZone"
>
  <LabelLine
    htmlFor="test"
  />
  <div
    className="DropZone__content"
    id="test"
    onClick={[Function]}
    onDragEnter={[Function]}
    onDragLeave={[Function]}
    onDragOver={[Function]}
    onDrop={[Function]}
    role="button"
    tabIndex={0}
  >
    <input
      onChange={[Function]}
      style={
        Object {
          "display": "none",
        }
      }
      type="file"
    />
    <div
      className="DropZone__uploaded"
    >
      <div>
        "test-file.txt" attached.
      </div>
      <SvgIcon
        className="DropZone__icon DropZone__icon--uploaded"
        icon={
          Object {
            "icon": Array [],
            "iconName": "file-alt",
            "prefix": "far",
          }
        }
      />
      <div
        className="DropZone__hint"
      >
        Uploaded hint
      </div>
    </div>
  </div>
  <FieldError />
</div>
`;
